add broadcasters

matgabriel opened the pull request:
There are probably some API details to improve

pdaures commented on mux/broadcaster.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// Register a new channel to receive broadcasts
+func (b *Broadcaster) Register(consumerName string, newch chan<- interface{}) {
```
Good job :)
Question: consumerName is only used to log error messages on backpressure, am I correct?

matgabriel replied:
yes, it is used only to identify the consumer that is dropping messages on backpressure

pdaures replied:
:+1:

matgabriel replied:
I modified a bit the API to pass the backpressure behavior on subscription, so each subscriber can have a custom behavior, and there is not need to pass the name anymore

pdaures commented on mux/broadcaster.go:
```diff
@@ -0,0 +1,98 @@
+		select {
+		case r, ok := <-b.reg:
+			if ok {
+				b.outputs[r.consumer.channel] = r.consumer.name
```
I didn't know we can use channel as key in a map!

pdaures commented on mux/broadcaster.go:
```diff
@@ -0,0 +1,98 @@
+// onBackPressureState is an action to execute when messages are dropped on back pressure (typically logging), it can be nil
+func NewNonBlockingBroadcaster(buflen int, onBackPressure func(consumerName string, value interface{})) *Broadcaster {
+	b := &Broadcaster{
+		input:   make(chan interface{}, buflen),
```
buflen would be bufLen

matgabriel replied:
yep

pdaures commented on mux/broadcaster_test.go:
```diff
@@ -0,0 +1,73 @@
+	"time"
+)
+
+const numberOfMessagesSent = 20
```
If we want to add tests in this file, we're going to have some problems because we have global values here.
I think we would be better to move the declaration of `numberOfMessageSent`, `blockingClientChan` and `finished`  inside TestBackpressureOnBroadcaster, and pass `finished` to `consume` as argument

matgabriel replied:
yep

pdaures commented on mux/broadcaster_test.go:
```diff
@@ -0,0 +1,73 @@
+	blockingChan := make(chan interface{}, 10)
+	consumeAndBlock(5, blockingChan)
+
+	normalChan := make(chan interface{}, numberOfMessagesSent+1)
```
maybe rename normalChan to nonBlockingChan?

matgabriel replied:
yes

pdaures commented on mux/state_test.go:
```diff
@@ -0,0 +1,162 @@
+	t.Log(fmt.Sprintf("backpressure count = %d", backpressureCount))
+}
+
+func TestFullStateSentToSubscriber(t *testing.T) {
```
:+1: very readable test

matgabriel commented:
I modified the run method in broadcaster: the idea is that when there are no subscribers, then we should not read from the input. This means that if we configure the input with a buffer size of 0, the producer should be blocked on the submit method if there are no subscribers.

This makes sense if the producer is reading from a queue (e.g. AMQP), and we do not want to consume to the queue when there are no subscribers.
I also added a method submitNonBlocking that will drop messages on backpressure.

next step: adding an optional behavior after the message was broadcasted (e.g. acknowledge the message)

pdaures commented:
J'ai fait des tests de perf en utilisant le broadcaster pour le streaming GRPC.

Avec 5 subscribers, on arrive à ~15k messages par seconde, ce qui est pas trop mal.
Avec 10 subscribers, on arrive à 10k messages par sec
à 100 subscribers, on est à 1.5k messages par seconde.

Pour le moment, ça semble assez rapide pour nos use cases. Mais si ça devient un problème, on devra adapter le code pour utiliser une sorte de circular buffer où le writer peut écrire sans s'occuper de chaque channel de reader.
